src/app/3d/qgs3dmaptoolpointcloudchangeattribute.h

```diff
@@ -66,32 +33,19 @@ class Qgs3DMapToolPointCloudChangeAttribute : public Qgs3DMapTool
     Qgs3DMapToolPointCloudChangeAttribute( Qgs3DMapCanvas *canvas );
     ~Qgs3DMapToolPointCloudChangeAttribute() override;
 
-    void mousePressEvent( QMouseEvent *event ) override;
-    void mouseReleaseEvent( QMouseEvent *event ) override;
-    void mouseMoveEvent( QMouseEvent *event ) override;
-    void keyPressEvent( QKeyEvent *event ) override;
-
-    void activate() override;
-    void deactivate() override;
-
-    bool allowsCameraControls() const override { return false; }
-
+    //! Sets attribute name which will be edited
     void setAttribute( const QString &attribute );
+    //! Sets attribute value to which it will be set
     void setNewValue( double value );
 
-  private:
-    void run();
-    void restart();
-    QgsPoint screenPointToMap( const QPoint &pos ) const;
-    SelectedPoints searchPoints( QgsPointCloudLayer *layer, const QgsGeos &searchPolygon ) const;
-    QVector<int> selectedPointsInNode( const QgsGeos &searchPolygon, const QgsPointCloudNodeId &n, const MapToPixel3D &mapToPixel3D, QgsPointCloudLayer *layer ) const;
-    static QgsGeometry box3DToPolygonInScreenSpace( const QgsBox3D &box, const MapToPixel3D &mapToPixel3D );
+  protected:
+    //! Calculate selection and edit set attribute to new value
+    virtual void run();
+    //! Clear selection
+    virtual void restart();
```

Comment:
do run() and restart() need to be virtual functions? they do not seem to be called from the base class - so they could be only defined as normal functions in subclasses